### PR TITLE
refactor(notifications): Updating notification logic

### PR DESCRIPTION
**Pull request by jeyrschabu**
- remove notificationConfiguration.required field
- remove notificationCount
- added a notification queue
- misc refactor for items no longer fitting the overall architecture

### CODE REVIEW

**Comment by jeyrschabu** [comment]
Notifications as currently designed would result in:

- Published notification events for resources not being notified on
- Too frequent notifications for small sets of resources to  an owner
- Separate notifications to the same user for different namespaces i.e `aws:servergroup:test:us-east-1` vs `aws:servergroup:test:us-west-2`

This PR solves the aforementioned problems by:
- Ensuring no events is published for resources not being notified on (amis & snapshots)
- Scheduling notifications to ensure more resources are packaged a notification to be sent to an owner
- Grouping notifications by resource type, ensuring resources belonging to different account/region pairs can be contained in the same notification

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/config/SwabbieProperties.kt [hunk 1]
```diff
@@ -69,25 +69,11 @@ class CloudProviderConfiguration {
 }
 
 /**
- * Notifications have both an [enabled] and a [required] configuration.
```
Comment (jeyrschabu) [inline swabbie-core/src/main/kotlin/com/netflix/spinnaker/config/SwabbieProperties.kt]: No longer applicable

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/AbstractResourceTypeHandler.kt [hunk 2]
```diff
@@ -108,10 +102,50 @@ abstract class AbstractResourceTypeHandler<T : Resource>(
   }
 
   /**
-   * Notifies resource owners
+   * Notifies for everything in resourceRepository.getMarkedResources() that doesn't have notificationInfo
    */
   override fun notify(workConfiguration: WorkConfiguration) {
```
Comment (jeyrschabu) [inline swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/AbstractResourceTypeHandler.kt]: This now schedules notifications when enabled, otherwise update the resource notification info so it can be ready for deletion. AMIs and Snapshots would fall in this category

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/AbstractResourceTypeHandler.kt [hunk 3]
```diff
@@ -582,122 +615,46 @@ abstract class AbstractResourceTypeHandler<T : Resource>(
...
+      }
+    val maxItemsToProcess = min(size, getMaxItemsProcessedPerCycle(workConfiguration))
+    return subList(0, maxItemsToProcess)
       .filter {
         @Suppress("UNCHECKED_CAST")
         !shouldExcludeResource(it.resource as T, workConfiguration, optedOutResourceStates, Action.NOTIFY)
       }
-      .groupBy { it.grouping?.value }
-
-    groupedResourcesToNotify.forEach { grouping, resources ->
```
Comment (jeyrschabu) [inline swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/AbstractResourceTypeHandler.kt]: Moved relevant bits `EchoNotifier`.

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/discovery/DiscoveryAware.kt [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.swabbie.discovery
+
+import com.netflix.spinnaker.kork.eureka.RemoteStatusChangedEvent
+import java.util.concurrent.atomic.AtomicBoolean
+
+open class DiscoveryAware : DiscoveryActivated {
```
Comment (jeyrschabu) [inline swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/discovery/DiscoveryAware.kt]: We have a few places where this is needed. Lifting this in a class so it can be used by any scheduler.

File: swabbie-echo/src/main/kotlin/com/netflix/spinnaker/swabbie/echo/EchoNotifier.kt [hunk 5]
```diff
@@ -16,23 +16,78 @@
...
+
+    throw UnsupportedOperationException("Notification Type not supported in $workConfiguration")
+  }
+
+  private fun String.unCamelCase(): String =
+    split("(?=[A-Z])".toRegex()).joinToString(" ").toLowerCase()
+
+  private fun notificationContext(
+    recipient: String,
+    resources: List<Pair<MarkedResource, WorkConfiguration>>,
```
Comment (jeyrschabu) [inline swabbie-echo/src/main/kotlin/com/netflix/spinnaker/swabbie/echo/EchoNotifier.kt]: This will require the echo notification template for swabbie to be updated to use this model. It makes determining metadata/configuration associated with a marked resource. We were previously passing just one because notifications were scoped to a specific namespace.  (cc: @aravindmd )

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/config/SwabbieProperties.kt (hunk 1 above)
Comment (emjburns) [reply swabbie-core/src/main/kotlin/com/netflix/spinnaker/config/SwabbieProperties.kt]: Thank goodness 😆

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/discovery/DiscoveryAware.kt (hunk 4 above)
Comment (emjburns) [reply swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/discovery/DiscoveryAware.kt]: What if we just added this up/down feature to `DiscoveryActivated`? I think that any class that is `DiscoveryActivated` wouldn't mind having this up atomic boolean, and that avoids another class to pick from.

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+  /**
+   * @return Marked resources to notify on
+   */
+  private fun groupRelatedResourcesThenNotify(markedResources: List<MarkedResource>) {
+    val now = Instant.now(clock)
+    val tasks = mutableListOf<NotificationTask>()
+    do {
+      val notificationTask: NotificationTask? = notificationQueue.pop()
+      if (notificationTask == null) {
```
Comment (emjburns) [inline swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt]: Could we just check `notificationQueue.isEmpty()` here?

Comment (emjburns) [reply swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt]: I guess, what's the case where the condition in this `do/while` loop would let this run but then there wouldn't be anything to notify? Are we locking on something else that would allow this to happen?

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+      return
+    }
+
+    lockingService.acquireLock(lockOptions) {
+      val resources = resourceTrackingRepository.getMarkedResources()
+        .filter {
+          it.notificationInfo == null
+        }
+
+      groupRelatedResourcesThenNotify(resources)
```
Comment (emjburns) [inline swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt]: I'd prefer to split this function out into several functions, but you don't have to do that. In general, I prefer things that are structured like: 
```
bigClassMethod() {
  val resultOne = stepOne()
  val resultTwo = stepTwo(resultOne)
  stepThree(resultTwo)
}
```
I find this easier to figure out where the work is in the class, and avoids the problem of "oh this method looks harmless ... three levels down.. oh wait we're doing a lot in here".

When looking at this function I would expect that first resources are grouped (in one method), and then those notifications are sent (in another method). That should be easier to debug and test as well.

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/config/SwabbieProperties.kt [hunk 8]
```diff
@@ -69,25 +69,11 @@ class CloudProviderConfiguration {
...
- *
- *  TODO EB: Remove the hard dependency on notifications for all resources.
- *   Instead, make this configurable in a logical and flexible way.
+ *  [enabled] on/off flag for notifications
  */
 open class NotificationConfiguration(
   var enabled: Boolean = false,
-  var required: Boolean = true,
-  var types: MutableList<String> = mutableListOf("Email"),
+  var types: List<String> = listOf("Email"),
```
Comment (emjburns) [inline swabbie-core/src/main/kotlin/com/netflix/spinnaker/config/SwabbieProperties.kt]: nit: it's unusual that email is capitalized. I'd change to lowercase

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationQueue.kt [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.swabbie.notifications
+
+/**
+ * A queue for storing notification tasks
+ */
+interface NotificationQueue {
+  fun add(notificationTask: NotificationSender.NotificationTask)
```
Comment (emjburns) [inline swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationQueue.kt]: I like it when you add an import for the class, so that it shows up as `notificationTask: NotificationTask`. `option + click` on `NotificationSender` in intellij should bring up that option. It's a little cleaner to read.

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+        }.groupBy {
+          it.resourceOwner
+        }
+
+      byResourceOwner.forEach { (owner, list) ->
+        val resourcesAndConfigurations: List<Pair<MarkedResource, WorkConfiguration>> = list.toResourceAndConfigurationPairs()
+        val result = notifier.notify(
+          Notifier.Envelope(
+            recipient = owner,
+            resources = resourcesAndConfigurations
```
Comment (emjburns) [inline swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt]: What's the value in passing the whole work configuration here? Does the notification system need everything? If not maybe we could pass a lot less information.

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+                resourceTrackingRepository.upsert(it)
+                applicationEventPublisher.publishEvent(OwnerNotifiedEvent(it, workConfiguration))
+              }
+
+            registry.counter(notificationsId.withTags("result", "success")).increment()
+          } else {
+            // unmark this resource & let it go through the mark cycle later
+            registry.counter(notificationsId.withTags("result", "failure")).increment()
+            resourceTrackingRepository.remove(resource)
+            applicationEventPublisher.publishEvent(UnMarkResourceEvent(resource, workConfiguration))
```
Comment (emjburns) [inline swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt]: I think this should be a different type of event, so that we can log metrics independently for things that fail to notify. I don't want to lump everything into the "unmark" bucket.

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+          it.notificationInfo == null
+        }
+
+      groupRelatedResourcesThenNotify(resources)
+    }
+  }
+
+  /**
+   * @return Marked resources to notify on
+   */
```
Comment (emjburns) [inline swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt]: I'd like a method level comment here explaining why we need to group resources and then how we're grouping and notifying. I find this method a little hard to follow.

File: swabbie-core/src/test/kotlin/com/netflix/spinnaker/swabbie/ResourceTypeHandlerTest.kt [hunk 13]
```diff
@@ -652,84 +650,73 @@ object ResourceTypeHandlerTest {
...
+
+    whenever(
+      notifier.notify(any())
+    ) doReturn NotificationResult(markedResource.resourceOwner, EMAIL, success = true)
+
+    defaultHandler.setCandidates(mutableListOf(defaultResource))
+    defaultHandler.setRules(defaultRules)
+
     defaultHandler.notify(configuration)
-    verify(notifier, times(1)).notify(any(), any(), any())
```
Comment (emjburns) [inline swabbie-core/src/test/kotlin/com/netflix/spinnaker/swabbie/ResourceTypeHandlerTest.kt]: Do you want to verify or assert anything in this test? FYI we're using Strikt in `keel` (http://strikt.io/) and it's really good for asserting simple things. I think I already added it to some of the swabbie tests.

File: swabbie-core/src/test/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSenderTest.kt [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+  }
+
+  @Test
+  fun `should not notify if there is nothing in the notification queue`() {
+    whenever(
+      resourceRepository.getMarkedResources()
+    ) doReturn listOf(
+      createMarkedResource(workConfiguration = workConfiguration2, id = "3", owner = "test@netflix.com")
+    )
+
```
Comment (emjburns) [inline swabbie-core/src/test/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSenderTest.kt]: Don't you actually need to call the notifier here?

File: swabbie-core/src/test/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSenderTest.kt [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+      verify(notifier, times(1)).notify(capture())
+      val passedResources = firstValue.resources.map { it.first }
+      expectThat(passedResources)
+        .isNotEmpty()
+        .isEqualTo(
+          listOf(resource1, resource2)
+        )
+    }
+
+    // Should publish events for resource1 and resource2 since they match the resource type
```
Comment (emjburns) [inline swabbie-core/src/test/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSenderTest.kt]: Maybe this test would be better if it was split out into two tests? Like separately test that notification info is populated correctly, and if type doesn't match then the resource is not notified?

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt (hunk 10 above)
Comment (emjburns) [reply swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt]: Ahh I didn't realize `Envelope` was only going to the `EchoNotifier`. To me envelope has the specific connotation of being used in a request/response body to wrap things. There's probably a clearer name for this.

File: swabbie-echo/src/main/kotlin/com/netflix/spinnaker/swabbie/echo/EchoNotifier.kt [hunk 16]
```diff
@@ -16,23 +16,78 @@
...
+            return NotificationResult(recipient, NotificationType.EMAIL, success = false)
+          }
+        }
+      }
+    }
+
+    throw UnsupportedOperationException("Notification Type not supported in $workConfiguration")
+  }
+
+  private fun String.unCamelCase(): String =
```
Comment (emjburns) [inline swabbie-echo/src/main/kotlin/com/netflix/spinnaker/swabbie/echo/EchoNotifier.kt]: Is there a commons lang function that already does this?

Comment (jeyrschabu) [reply swabbie-echo/src/main/kotlin/com/netflix/spinnaker/swabbie/echo/EchoNotifier.kt]: See TODO comment. I'd prefer changing resource types into something that can provide a human readable name rather than trying to normalize this way. Leaving this as-is for now. I will address this in a different refactor.

File: swabbie-core/src/test/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSenderTest.kt (hunk 15 above)
Comment (jeyrschabu) [reply swabbie-core/src/test/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSenderTest.kt]: There are still a couple of scenarios I'd like tested in this class. I will add separate tests.

File: swabbie-core/src/test/kotlin/com/netflix/spinnaker/swabbie/ResourceTypeHandlerTest.kt (hunk 13 above)
Comment (jeyrschabu) [reply swabbie-core/src/test/kotlin/com/netflix/spinnaker/swabbie/ResourceTypeHandlerTest.kt]: I will take a look. `Strikt` is used in a few places including in `NotificationSenderTest`

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt (hunk 11 above)
Comment (jeyrschabu) [reply swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt]: There is a counter right above tracking notification failures.

Thinking about this more, the resource would get picked up when we notify on the same resource type. So no need to remove and unmark. It could use a debug level log here as well.

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/AbstractResourceTypeHandler.kt (hunk 2 above)
Comment (aravindmd) [reply swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/AbstractResourceTypeHandler.kt]: do we have any tests to validate this for AMI's and snapshots ?

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/discovery/DiscoveryAware.kt (hunk 4 above)
Comment (jeyrschabu) [reply swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/discovery/DiscoveryAware.kt]: Fine by me, I suppose that can be implemented in `DiscoveryActivated`.

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt (hunk 7 above)
Comment (jeyrschabu) [reply swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt]: This is good feedback!

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt (hunk 10 above)
Comment (jeyrschabu) [reply swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt]: It doesn't have the same connotation to me. I am not married to that name so what's your suggestion?

File: swabbie-core/src/test/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSenderTest.kt (hunk 14 above)
Comment (jeyrschabu) [reply swabbie-core/src/test/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSenderTest.kt]: Planning on revisiting this test

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/AbstractResourceTypeHandler.kt (hunk 2 above)
Comment (jeyrschabu) [reply swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/AbstractResourceTypeHandler.kt]: Please elaborate

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/config/SwabbieProperties.kt (hunk 8 above)
Comment (jeyrschabu) [reply swabbie-core/src/main/kotlin/com/netflix/spinnaker/config/SwabbieProperties.kt]: Whatever fine!

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationQueue.kt (hunk 9 above)
Comment (jeyrschabu) [reply swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationQueue.kt]: if it were a top level class yes. I prefer closer to a fully qualified name for internal classes.

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+      val byResourceOwner = markedResources
+        .filter {
+          it.resourceType == task.resourceType
+        }.groupBy {
+          it.resourceOwner
+        }
+
+      byResourceOwner.forEach { (owner, list) ->
+        val resourcesAndConfigurations: List<Pair<MarkedResource, WorkConfiguration>> = list.toResourceAndConfigurationPairs()
+        val result = notifier.notify(
```
Comment (aravindmd) [inline swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt]: would we need to check the violations again before sending a notify event ? would there be a case where the violations could have changed when the resource is picked up for notification ?

File: swabbie-echo/src/main/kotlin/com/netflix/spinnaker/swabbie/echo/EchoNotifier.kt (hunk 5 above)
Comment (aravindmd) [reply swabbie-echo/src/main/kotlin/com/netflix/spinnaker/swabbie/echo/EchoNotifier.kt]: I will update the template once this is merged.

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt (hunk 6 above)
Comment (jeyrschabu) [reply swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt]: A `do while` construct allows allows to check for that condition once. Otherwise you would have to check if the queue is empty, then proceed to pop items off of it and continue to check if the queue is empty.

Comment (jeyrschabu) [reply swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt]: Not sure I follow the second part of the comment: `Are we locking on something else that would allow this to happen?`

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt (hunk 17 above)
Comment (jeyrschabu) [reply swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt]: Resources are fetched lazily by the notification sender + there is the same locking mechanism we have elsewhere

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/AbstractResourceTypeHandler.kt (hunk 2 above)
Comment (aravindmd) [reply swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/AbstractResourceTypeHandler.kt]: Ignore , i saw that you have the tests based on `resourceType` in `NotificationSenderTest`

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationQueue.kt (hunk 9 above)
Comment (emjburns) [reply swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationQueue.kt]: I think it's weird that `NotificationTask` is an internal class and yet it's referenced by two other files. Maybe best to move to its own file?

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt (hunk 6 above)
Comment (emjburns) [reply swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt]: I was confused reading this at first because I didn't see the `do/while` and I didn't go back and update my comment. What if we just had a `popAll` method on the queue that returned everything?

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt (hunk 10 above)
Comment (emjburns) [reply swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt]: oof hit me back with the "what's your suggestion" 😆 good one! I'll think about it and try to offer one on my next pass of PR review.

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt (hunk 11 above)
Comment (emjburns) [reply swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt]: ahh I read so well. I totally see that now :)

**Comment by jeyrschabu** [comment]
I was already in the code so I cut this guy to cover the echo side once this is merged:

https://github.com/spinnaker/echo/pull/671

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/MetricsSupport.kt [hunk 18]
```diff
@@ -23,6 +23,7 @@ import com.netflix.spinnaker.swabbie.events.Action
 import com.netflix.spinnaker.swabbie.model.WorkConfiguration
 import java.util.concurrent.atomic.AtomicInteger
 
+// TODO: Jeyrs - refactor me I hate this!!!!
```
Comment (emjburns) [inline swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/MetricsSupport.kt]: 😆

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/discovery/DiscoveryActivated.kt [hunk 19]
```diff
@@ -4,15 +4,25 @@ import com.netflix.appinfo.InstanceInfo
...
+import java.util.concurrent.atomic.AtomicBoolean
 
 /**
  * A component that starts doing something when the instance is up in discovery
  * and stops doing that thing when it goes down.
  */
-interface DiscoveryActivated : ApplicationListener<RemoteStatusChangedEvent> {
+open class DiscoveryActivated : ApplicationListener<RemoteStatusChangedEvent> {
+  private val up = AtomicBoolean()
+  open fun onDiscoveryUpCallback(event: RemoteStatusChangedEvent) {
```
Comment (emjburns) [inline swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/discovery/DiscoveryActivated.kt]: I think these should remain callbacks with no body, and the up/down logic should go in the `onApplicationEvent` function. That way, you can override these to provide specific logic without messing up the up/down indication or having to call `super`.

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/model/Resource.kt [hunk 20]
```diff
@@ -119,6 +120,22 @@ interface Identifiable : Named {
   val resourceType: String
   val cloudProvider: String
   val grouping: Grouping?
+
+  /**
+   * Returns the opt-out url for this resource
+   * This is used for notifications
+   */
+  fun optOutUrl(workConfiguration: WorkConfiguration): String {
```
Comment (emjburns) [inline swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/model/Resource.kt]: Does this have any impact on json serialization/deserialization? aka will making changes here lead to problems reading from the current database when we roll it out?

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt [hunk 21]
```diff
@@ -0,0 +1,220 @@
...
+   * The frequency of this scheduled action is cron based and configurable
+   * Defaults to every day at 9am.
+   */
+  @Scheduled(cron = "\${swabbie.notification.cron.schedule:0 0 9 1/1 * ?}")
+  fun sendNotifications() {
+    if (!isUp()) {
+      return
+    }
+
+    lockingService.acquireLock(lockOptions) {
```
Comment (emjburns) [inline swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt]: This is so much clearer to me. Thanks!

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/work/WorkQueueManager.kt [hunk 22]
```diff
@@ -48,37 +47,28 @@ class WorkQueueManager(
...
-        ensureLoadedCaches()
-        queue.refillOnEmpty()
-      }
-    }
-
-  override val onDiscoveryDownCallback: (event: RemoteStatusChangedEvent) -> Unit
-    get() = {
-      up.set(false)
+  override fun onDiscoveryUpCallback(event: RemoteStatusChangedEvent) {
+    if (isEnabled()) {
```
Comment (emjburns) [inline swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/work/WorkQueueManager.kt]: as the code is now, this `override` makes it so `DiscoveryActivated` never does the up/down business. see my earlier comment in that class.

File: swabbie-core/src/test/kotlin/com/netflix/spinnaker/swabbie/ResourceTypeHandlerTest.kt [hunk 23]
```diff
@@ -652,84 +653,76 @@ object ResourceTypeHandlerTest {
...
+    ) doReturn listOf(markedResource)
+
+    defaultHandler.setCandidates(mutableListOf(defaultResource))
+    defaultHandler.setRules(defaultRules)
+
     defaultHandler.notify(configuration)
-    verify(notifier, times(0)).notify(any(), any(), any())
-    verify(resourceRepository, times(1)).upsert(updatedMarkedResource)
+
+    with(markedResource.notificationInfo) {
```
Comment (emjburns) [inline swabbie-core/src/test/kotlin/com/netflix/spinnaker/swabbie/ResourceTypeHandlerTest.kt]: You can use an expect block in strikt, which i think will make this better. I don't want to have to assert that `this` is not null, reads kind of weird in the test.

You might just be able to do:
```kotlin
val notificationInfo = markedResource.notificationInfo
expect {
  that(notificationInfo).isNotNull()
  ...
  that(notificationInfo.NotificationStamp).isNull()
}
```

File: swabbie-redis/src/test/kotlin/com/netflix/spinnaker/swabbie/redis/RedisNotificationQueueTest.kt [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.TestInstance
+import redis.clients.jedis.JedisPool
+import strikt.api.expectThat
+import strikt.assertions.isEqualTo
+import strikt.assertions.isFalse
+import strikt.assertions.isTrue
+import strikt.assertions.size
+
```
Comment (emjburns) [inline swabbie-redis/src/test/kotlin/com/netflix/spinnaker/swabbie/redis/RedisNotificationQueueTest.kt]: Something that Fletch does that I really like is he sets up a base test class (for example, [ResourceRepositoryTests](https://github.com/spinnaker/keel/blob/master/keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/ResourceRepositoryTests.kt)) and then each repo type (in memory, sql, redis, etc) extends that test to provide whatever initialization logic they need [InMemory](https://github.com/spinnaker/keel/blob/master/keel-core/src/test/kotlin/com/netflix/spinnaker/keel/persistence/memory/InMemoryResourceRepositoryTest.kt), [sql](https://github.com/spinnaker/keel/blob/master/keel-sql/src/test/kotlin/com/netflix/spinnaker/keel/sql/SqlResourceRepositoryTests.kt)).

This makes it so that you're _sure_ the in memory and real implementations work the same. Something to keep in mind for the future!

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/discovery/DiscoveryActivated.kt (hunk 19 above)
Comment (jeyrschabu) [reply swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/discovery/DiscoveryActivated.kt]: Yes, the up/down can go in the onApplication body but I prefer these as overridable functions rather than properties though. The end result would be the same but reads better as functions.

File: swabbie-redis/src/test/kotlin/com/netflix/spinnaker/swabbie/redis/RedisNotificationQueueTest.kt (hunk 24 above)
Comment (jeyrschabu) [reply swabbie-redis/src/test/kotlin/com/netflix/spinnaker/swabbie/redis/RedisNotificationQueueTest.kt]: Got plans to write a proper `tck` at some point.

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/model/Resource.kt (hunk 20 above)
Comment (jeyrschabu) [reply swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/model/Resource.kt]: I dont believe so. Will test that out.

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/discovery/DiscoveryActivated.kt (hunk 19 above)
Comment (jeyrschabu) [reply swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/discovery/DiscoveryActivated.kt]: grr can't keep that without changing a bunch other stuff. Will keep them as getters

**Comment by jeyrschabu** [comment]
Ok will update this with the feedback re:discoveryActivated and mergeroni. Likely to tune things out depending how testing goes.
